### ISSUE TITLE
mtime of taskchampion.sqlite3 got changed when calling `task list`

### ISSUE THREAD

**Issue by chenyuntc** [GothenburgBitFactory/taskwarrior#3418]
When I run `task list` the ` taskchampion.sqlite3`'s mtime changed.

This is inconvenience as I use the mtime to sync between workstations.

So is there a way to turn off the changes on the sqlite file when operations has no side-effect (i.e. list, filter)?

**Comment by LeGuipo** [comment]
I confirm. Every time taskwarrior fetches some data, the file seems updated. I also suspect this is the cause of the noticeable delays for taskwarrior to display results.

**Comment by djmitche** [comment]
I don't think TW is acutally modifying the data, but it is opening the DB in r/w mode which, from some quick `strace` analysis, does perform a write of the first block.

So, we should find a way to open the DB in a read-only mode.

**Comment by djmitche** [comment]
#3753 got some way toward this, in the sense that it centralizes one of the two "automatic" changes to the task DB.

The problem is, just about every interesting command runs one of GC or handleRecurrence/handleUntil (I'll refer to these collectively as "updates" here). To support that naively, we need to open the DB in read-write mode for just about every invocation -- and in particular for reports, which is the slow part.

I can think of a few ways to fix this:

1. Open in read-only mode by default, but re-open in read-write mode if there's updates needed.
2. Always open in read-write mode unless some rc.read-only=1 is set, in which case open read-only but don't do any updates.
3. Probabalistically open in read-write mode to perform updates.

The first option is more difficult than it seems, as the GC is performed in TaskChampion and does not currently provide a way to "try" with an interpretable error when updates are actually required.

The second option is pretty bad UX, as it basically means if things are slow you need to distribute `rc.read-only=1` all over the place.

The third option would help with performance, but only sometimes, and introduces non-determinacy into the application, making understanding and debugging more difficult.

Anyone have other ideas?

**Comment by djmitche** [comment]
On reflection, a fourth option similar to 2 is to open in read-write mode unless `rc.gc=0`. That's an existing config option that disables garbage collection, and is already suggested for use from scripts, etc.

**Comment by me-and** [comment]
I'm following this ticket as a specific aspect of #3329, which is currently the thing keeping me on Taskwarrior 2. Certainly for me – and I expect a lot of people who routinely have relatively large task lists – I have `gc=0` and `recurrence=0` in my config, and just run commands to override those every so often.

Personally, I'd rather have an option for Taskwarrior to just not have side-effects: it shouldn't collect garbage unless I ask it to, it shouldn't create recurring tasks unless I ask it to, it shouldn't delete tasks past their "until" date unless I ask it to. That's probably too big a change to make the default (much as I'd like that), but making it an option seems like it might solve a lot of problems…

**Comment by djmitche** [comment]
I agree -- I was surprised to find that it would do the recurrence/until thing autonomously! But as you say, that's a big change. And, I think it's already an option in as much as `rc.gc=0` turns these things off. Or, I thought it did (#3772)!

**Comment by meonkeys** [comment]
I'm using `gc=0` in my taskrc and am running `task rc.gc=1` daily with anacron. Is this a good idea? If yes, would it also make sense for taskwarrior itself to do a daily garbage collection? If yes, why not weekly or monthly?

I'm naive as to what all exactly is happening when garbage collection occurs. I'll keep searching docs and code and try to figure it out.

**Comment by smemsh** [comment]
I run with gc off myself, but don't run it every day, just manually every "once in a while" or on demand.  There's no harm in having gc off.  It only prunes deleted/purged/done stuff and shifts the remaining ones' ID numbers to start from 1.  This way the IDs of active tasks are always low integers (less typing).  The UUID of course stays the same.
